File: common/src/main/java/generations/gg/generations/core/generationscore/client/screen/SubTexture.java
package generations.gg.generations.core.generationscore.client.screen;

import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.resources.ResourceLocation;

/**
 * Used for working with Sprite Sheets. Allows for "mapping" coordinates inside the image to something which can be drawn.
 */
public class SubTexture {

    protected final ResourceLocation sheet;
    protected final int u;
    protected final int v;
    protected final int uWidth;
    protected final int vHeight;
    protected final int width;
    protected final int height;
    protected final int sheetScale;

    public SubTexture(int u, int v, int width, int height, ResourceLocation sheet) {
        this(u, v, width, height, sheet, 1024);
    }

    public SubTexture(int u, int v, int width, int height, ResourceLocation sheet, int scale) {
        this.u = u;
        this.v = v;
        this.uWidth = width;
        this.vHeight = height;
        this.width = width;
        this.height = height;
        this.sheet = sheet;
        this.sheetScale = scale;
    }

    public void render(GuiGraphics stack, int x, int y, boolean centered) {
        renderInner(stack, centered ? x - this.width / 2 : x, y, this.sheetScale, this.sheetScale);
    }

    public void renderStretched(GuiGraphics stack, int x, int y, float widthMultiplier, float heightMultiplier) {
        ScreenUtils.drawTexture(stack, sheet, x, y, (int) (this.width * widthMultiplier), (int) (this.height * heightMultiplier), this.u, this.v, this.uWidth, this.vHeight, this.sheetScale, this.sheetScale);
    }

    protected void renderInner(GuiGraphics stack, int x, int y, int texWidth, int texHeight) {
        ScreenUtils.drawTexture(stack, sheet, x, y, this.width, this.height, this.u, this.v, this.uWidth, this.vHeight, texWidth, texHeight);
    }

    public void render(GuiGraphics stack, int x, int y, int width, int height) {
        ScreenUtils.drawTexture(stack, sheet, x, y, width, height, this.u, this.v, this.uWidth, this.vHeight, this.sheetScale, this.sheetScale);
    }

    public void render(GuiGraphics stack, int x, int y, int color) {
        render(stack, x, y, width, height, color);
    }

    public void render(GuiGraphics stack, int x, int y, int width, int height, int color) {
        float startAlpha = (float)(color >> 24 & 0xFF) / 255.0f;
        float startRed = (float)(color >> 16 & 0xFF) / 255.0f;
        float startGreen = (float)(color >> 8 & 0xFF) / 255.0f;
        float startBlue = (float)(color & 0xFF) / 255.0f;

        RenderSystem.setShaderColor(startRed, startGreen, startBlue, startAlpha);
        render(stack, x, y, width, height);
        RenderSystem.setShaderColor(1.0f, 1.0f, 1.0f, 1.0f);
    }

    public void renderAnchored(GuiGraphics stack, int x, int y, Anchor anchor) {
        ScreenUtils.drawAnchoredTexture(stack, sheet, x, y, width, height, this.u, this.v, this.uWidth, this.vHeight, this.sheetScale, this.sheetScale, anchor);
    }

    public void renderAnchoredStretched(GuiGraphics stack, int x, int y, float widthMultiplier, float heightMultiplier, Anchor anchor) {
        ScreenUtils.drawAnchoredStretchedTexture(stack, sheet, x, y, width, height, this.u, this.v, this.uWidth, this.vHeight, this.sheetScale, this.sheetScale, widthMultiplier, heightMultiplier, anchor);
    }

    public static final SubTexture BLANK = new SubTexture(0, 0, 0, 0, null) {
        public void render(GuiGraphics stack, int x, int y, boolean centered) {}
        public void renderStretched(GuiGraphics stack, int x, int y, float widthMultiplier, float heightMultiplier) {}
        protected void renderInner(GuiGraphics stack, int x, int y, int texWidth, int texHeight) {}
        public void renderAnchored(GuiGraphics stack, int x, int y, Anchor anchor) {}
        public void renderAnchoredStretched(GuiGraphics stack, int x, int y, float widthMultiplier, float heightMultiplier, Anchor anchor) {}
    };

    public void render(GuiGraphics poseStack, int x, int y) {
        this.render(poseStack, x, y, false);
    }
}

File: src/common/main/kotlin/generations/gg/generations/core/generationscore/common/client/screen/SubTexture.kt
package generations.gg.generations.core.generationscore.common.client.screen

import com.mojang.blaze3d.systems.RenderSystem
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.renderer.texture.MissingTextureAtlasSprite
import net.minecraft.resources.ResourceLocation

/**
 * Used for working with Sprite Sheets. Allows for "mapping" coordinates inside the image to something which can be drawn.
 */
open class SubTexture @JvmOverloads constructor(
    protected val u: Int,
    protected val v: Int,
    protected val width: Int,
    protected val height: Int,
    protected val sheet: ResourceLocation,
    protected val sheetScale: Int = 1024
) {
    open fun render(graphics: GuiGraphics, x: Int, y: Int, centered: Boolean) {
        renderInner(graphics, if (centered) x - this.width / 2 else x, y, this.sheetScale, this.sheetScale)
    }

    open fun renderStretched(graphics: GuiGraphics, x: Int, y: Int, widthMultiplier: Float, heightMultiplier: Float) {
        ScreenUtils.drawTexture(
            graphics,
            sheet, x, y, (this.width * widthMultiplier).toInt(), (this.height * heightMultiplier).toInt(),
            u.toFloat(),
            v.toFloat(), this.width, this.height, this.sheetScale, this.sheetScale
        )
    }

    protected open fun renderInner(graphics: GuiGraphics, x: Int, y: Int, texWidth: Int, texHeight: Int) {
        ScreenUtils.drawTexture(
            graphics,
            sheet, x, y, this.width, this.height,
            u.toFloat(),
            v.toFloat(), this.width, this.height, texWidth, texHeight
        )
    }

    open fun render(graphics: GuiGraphics, x: Int, y: Int, width: Int, height: Int) {
        ScreenUtils.drawTexture(
            graphics,
            sheet, x, y, width, height,
            u.toFloat(),
            v.toFloat(), this.width, this.height, this.sheetScale, this.sheetScale
        )
    }

    fun render(graphics: GuiGraphics, x: Int, y: Int, color: Int) {
        render(graphics, x, y, width, height, color)
    }

    fun render(graphics: GuiGraphics, x: Int, y: Int, width: Int, height: Int, color: Int) {
        val startAlpha = (color shr 24 and 0xFF).toFloat() / 255.0f
        val startRed = (color shr 16 and 0xFF).toFloat() / 255.0f
        val startGreen = (color shr 8 and 0xFF).toFloat() / 255.0f
        val startBlue = (color and 0xFF).toFloat() / 255.0f

        RenderSystem.setShaderColor(startRed, startGreen, startBlue, startAlpha)
        render(graphics, x, y, width, height)
        RenderSystem.setShaderColor(1.0f, 1.0f, 1.0f, 1.0f)
    }

    open fun renderAnchored(graphics: GuiGraphics, x: Int, y: Int, anchor: Anchor) {
        ScreenUtils.drawAnchoredTexture(
            graphics,
            sheet, x, y,
            width,
            height, this.u, this.v, this.width, this.height, this.sheetScale, this.sheetScale, anchor
        )
    }

    open fun renderAnchoredStretched(
        graphics: GuiGraphics,
        x: Int,
        y: Int,
        widthMultiplier: Float,
        heightMultiplier: Float,
        anchor: Anchor
    ) {
        ScreenUtils.drawAnchoredStretchedTexture(
            graphics,
            sheet, x, y,
            width,
            height,
            this.u,
            this.v,
            this.width,
            this.height,
            this.sheetScale,
            this.sheetScale, widthMultiplier, heightMultiplier, anchor
        )
    }

    fun render(graphics: GuiGraphics, x: Int, y: Int) {
        this.render(graphics, x, y, false)
    }

    companion object {
        val BLANK: SubTexture = object : SubTexture(0, 0, 0, 0, MissingTextureAtlasSprite.getLocation()) {
            override fun render(graphics: GuiGraphics, x: Int, y: Int, centered: Boolean) {}
            override fun renderStretched(
                graphics: GuiGraphics,
                x: Int,
                y: Int,
                widthMultiplier: Float,
                heightMultiplier: Float
            ) {
            }

            override fun renderInner(graphics: GuiGraphics, x: Int, y: Int, texWidth: Int, texHeight: Int) {}
            override fun renderAnchored(graphics: GuiGraphics, x: Int, y: Int, anchor: Anchor) {}
            override fun renderAnchoredStretched(
                graphics: GuiGraphics,
                x: Int,
                y: Int,
                widthMultiplier: Float,
                heightMultiplier: Float,
                anchor: Anchor
            ) {
            }
        }
    }
}